src/CodeGen_LLVM.cpp

```diff
@@ -4722,6 +4706,92 @@ Value *CodeGen_LLVM::get_user_context() const {
     return ctx;
 }
 
+void CodeGen_LLVM::declare_intrin_overload(const std::string &name, const Type &ret_type, const std::string &impl_name, std::vector<Type> arg_types) {
+    llvm::Function *intrin = module->getFunction(impl_name);
+    if (!intrin) {
+        vector<llvm::Type *> llvm_arg_types(arg_types.size());
+        for (size_t i = 0; i < arg_types.size(); i++) {
+            llvm_arg_types[i] = llvm_type_of(arg_types[i]);
+        }
+
+        llvm::Type *llvm_ret_type = llvm_type_of(ret_type);
+        FunctionType *func_t = FunctionType::get(llvm_ret_type, llvm_arg_types, false);
+        intrin = llvm::Function::Create(func_t, llvm::Function::ExternalLinkage, impl_name, module.get());
+        intrin->setCallingConv(CallingConv::C);
+    }
+    intrinsics[name].emplace_back(ret_type, std::move(arg_types), intrin);
+}
+
+Value *CodeGen_LLVM::call_overloaded_intrin(const Type &result_type, const std::string &name, const std::vector<Expr> &args) {
+    auto impls_i = intrinsics.find(name);
+    if (impls_i == intrinsics.end()) {
+        debug(2) << "No intrinsic " << name << "\n";
+        return nullptr;
+    }
+
+    const Intrinsic *resolved = nullptr;
+    for (const Intrinsic &i : impls_i->second) {
+        if (i.arg_types.size() != args.size()) {
+            continue;
+        }
+
+        if (i.result_type.with_lanes(1) != result_type.with_lanes(1)) {
+            continue;
+        }
+
+        bool match = true;
+        for (int j = 0; j < (int)i.arg_types.size(); j++) {
+            if (i.arg_types[j].with_lanes(1) != args[j].type().with_lanes(1)) {
```

Comment:
We have Type::element_of for with_lanes(1). It only saves a single character, so if you think "with_lanes(1)" is clearer perhaps we should just globally replace element_of() with with_lanes(1), but that should be a separate PR.